### ISSUE TITLE
About the pad issues in the code

### ISSUE THREAD

**Issue by hyc9** [BELIEVEfxy/LightSANs#3]
Thank you for sharing the code! I noticed that there seems to be no proper masking of the pad in the sequence in your code. In the item embedding, pad_idx is set to 0, but after the init_weights operation, the embedding corresponding to the pad seems no longer to be a zero vector. 

I experimented with your code on my own dataset and found that the model's performance improved further after adding mask processing to the pad (I processed it in the attention of ItemToInterestAggregation). Even directly setting the embedding corresponding to the pad to zero can improve the model's performance. 

However, my experiment may not necessarily prove the necessity of handling pad in the code, or I may have missed the steps in your code for handling pad. If there's any misunstanding, please let me know, thanks a lot !

**Comment by zyx1017** [comment]
@hyc9 
大佬，对不起，我有个疑问：
在计算多头注意力得分时，为什么不需要像以前的做法一样，需要把之后的交互给mask掉？
比如以往的做法都会加上一个mask矩阵：attention_scores = attention_scores + attention_mask
而本代码见如下，并没有进行mask操作。对此感到疑问！希望能得到解答！！！
attention_scores = torch.matmul(query_layer, key_layer.transpose(-1, -2))
attention_scores = attention_scores / math.sqrt(self.attention_head_size)

\# normalize the attention scores to probabilities.
attention_probs = nn.Softmax(dim=-2)(attention_scores)
attention_probs = self.attn_dropout(attention_probs)
context_layer_item = torch.matmul(attention_probs, value_layer)

**Comment by hyc9** [comment]
> @hyc9 大佬，对不起，我有个疑问： 在计算多头注意力得分时，为什么不需要像以前的做法一样，需要把之后的交互给mask掉？ 比如以往的做法都会加上一个mask矩阵：attention_scores = attention_scores + attention_mask 而本代码见如下，并没有进行mask操作。对此感到疑问！希望能得到解答！！！ attention_scores = torch.matmul(query_layer, key_layer.transpose(-1, -2)) attention_scores = attention_scores / math.sqrt(self.attention_head_size)
> 
> normalize the attention scores to probabilities. attention_probs = nn.Softmax(dim=-2)(attention_scores) attention_probs = self.attn_dropout(attention_probs) context_layer_item = torch.matmul(attention_probs, value_layer)

如果是单向建模的自注意力序列模型，是需要mask掉之后的。但是这个模型里的k和v做了聚合处理，类似于“降低了序列长度“，每一个位置的向量都是看得到整个序列的，所以mask掉之后没有意义。这也是这个模型不能采用自回归的训练方式（即1,2,3,4->2,3,4,5)的理由

**Comment by zyx1017** [comment]
嗷嗷，了解了！谢谢

**Comment by zyx1017** [comment]
@hyc9 大佬，这里的代码所有的序列推荐的训练方式好像都是采用自回归的方式？

**Comment by hyc9** [comment]
> 

这里代码采用的应该是1,2,3,4->5这样的方式 + 序列数据增强吧

**Comment by hyc9** [comment]
> @hyc9 大佬，这里的代码所有的序列推荐的训练方式好像都是采用自回归的方式？

这里代码采用的应该是1,2,3,4->5这样的方式 + 序列数据增强吧

**Comment by zyx1017** [comment]
> > @hyc9 大佬，这里的代码所有的序列推荐的训练方式好像都是采用自回归的方式？
> 
> 这里代码采用的应该是1,2,3,4->5这样的方式 + 序列数据增强吧

他做数据增强操作的时候好像就是这么做的：[ [1,], [1,2,], [1,2,3,], [1,2,3,4,] ]  -> [2,3,4,5]
这不就类似自回归的训练方式吗？对比SASRec这个模型，与LightSANS的训练方式应该是一样的，但是SASRec在进行多头注意力的时候有进行mask操作。很奇怪
